rfcs/2021-12-20-10517-llvm-backend-for-vrl.md

```diff
@@ -0,0 +1,756 @@
+  call void @llvm.lifetime.end.p0i8(i64 40, i8* nonnull %58), !noalias !99611
+  %62 = getelementptr inbounds %116, %116* %7, i64 0, i32 0
+  %63 = load {}*, {}** %62, align 8, !noalias !99611
+  %64 = icmp eq {}* %63, null
+  %65 = bitcast {}* %63 to i8*
+  br i1 %64, label %71, label %66
+
+66:                                               ; preds = %51
+  %67 = getelementptr inbounds %116, %116* %7, i64 0, i32 1, i64 0
+  %68 = load i64, i64* %67, align 8, !noalias !99611
+  %69 = icmp eq i64 %68, 0
+  br i1 %69, label %71, label %70
+
+70:                                               ; preds = %66
+  call void @__rust_dealloc(i8* nonnull %65, i64 %68, i64 1) #104, !noalias !99608
+  br label %71
+
+71:                                               ; preds = %51, %66, %70
+  call void @llvm.lifetime.end.p0i8(i64 24, i8* nonnull %53), !noalias !99611
+  call void @llvm.lifetime.end.p0i8(i64 40, i8* nonnull %52), !noalias !99611
+  br label %47
+}
+```
+
+Note the batched stack allocations, inlining of function calls and consolidation
+of control flow.
+
+## Rationale
+
+As long as the single-core performance of VRL is the bottleneck of a topology,
+general performance improvements to VRL are extremely valuable as they equate to
+an equally sized performance improvement to the entire topology.
+
+We want to live up to the
+[performance guarantees](https://github.com/vectordotdev/vector/blob/f1404bea186ba83c4426a32bbef3f633c17cf4d2/website/cue/reference/remap/features/compilation.cue#L4-L8)
+outlined in VRL's list of features.
+
+Every VRL program benefits from the reduced runtime overhead, without us needing
+to optimize any specific use case. Consistent execution speed is important to
+build trust in the language.
+
+Being able to execute our log transformation DSL at speeds which would otherwise
+only be attainable by hand-writing Rust programs will strengthen a key value
+proposition of Vector: best-in-class performance.
+
+## Drawbacks
+
+By generating machine code via LLVM, we are no longer (largely) immune to memory
+violations. To reduce the error surface as much as possible, we employ industry
+practices like fuzz-testing VRL programs, static analysis via LLVM and rely on
+the Rust compiler for any non-trivial code fragments. That being said, memory
+safety is an illusion - with the difference that we are now exposing ourselves
+to being partly responsible for maintaining invariants instead of being able to
+defer to a third party for correctness. We still provide an inherently memory
+safe language to the user.
+
+Producing LLVM bitcode for Rust's `std` library is guarded behind the
+[`-Z build-std`](https://doc.rust-lang.org/cargo/reference/unstable.html#build-std)
+flag and only available on the nightly compiler toolchain. We need `std` to
+fully link our precompiled LLVM bitcode. It's possible to circumvent the nightly
+requirement by setting the `RUSTC_BOOTSTRAP=1` environment variable, such that
```

Comment:
The [documentation for this](https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html) mentions:

> This should never be used except when bootstrapping the [Rust] compiler.

Do we know what the implications are for enabling this flag? Should we investigate this before we consider this to be a viable solution?

Reply:
Do we need to circumvent the nightly requirement? Presumably it would still generate the same code?

Reply:
> Presumably it would still generate the same code?

I am not convinced that's a guarantee. Rust doesn't have a stable interface, so things might change between versions (although it's not very likely to happen without advanced warning), which would first show up in a nightly release.

Reply:
It sounds like the crux is that using `RUSTC_BOOTSTRAP=1` primarily removes the stability checks that prevent you from using nightly-only features on stable.  If we were worried about this, conceivably we could do something like a dummy project whose only purpose is to allow being compiled with the Rust stdlib so that we could generate the necessary LLVM bitcode for the stdlib functions?

Like, as long as we always use the same compiler for doing that step as we do for compiling VRL itself, and Vector, there should be no difference in codegen within a single build of Vector... obviously this may not hold between versions of Vector but I don't see where they would be a problem anyways.

Reply:
There's a big rabbit hole discussing that very question in https://github.com/rust-lang/cargo/issues/6627 and https://github.com/rust-lang/cargo/issues/7088.

We're exactly in the situation mentioned in https://github.com/rust-lang/cargo/issues/6627#issuecomment-460549562:

> Firefox doesn't want to use a "stable compiler", Firefox wants to use a "pinned nightly compiler" that happens to be pinned to exactly the same commit as stable release.
> [...]
> The problem is that `RUST_BOOTSTRAP` is the least painful way to do it, despite not being related to bootstrapping rustc.

By isolating the usage of the `RUSTC_BOOTSTRAP` to really building `std` only, the only thing we expose ourselves to is when Rust decides to revoke the ability to emit LLVM IR for `std` or Rust altogether. I think that's very unlikely and would be the least of our problems in that case. I think it's more reasonable to assume that at some point this will be possible in `stable`, allowing us to remove this isolated hack.

> we could do something like a dummy project whose only purpose is to allow being compiled with the Rust stdlib so that we could generate the necessary LLVM bitcode for the stdlib functions

@tobz yep, that's also what I would do. Took that approach in https://github.com/vectordotdev/vector/pull/10442. The [`std` crate](https://github.com/vectordotdev/vector/tree/ea68951b5949e77504fd8e5a829a7bcc722946b9/lib/vrl/compiler/src/precompiled/std) builds an empty [`lib.rs`](https://github.com/vectordotdev/vector/blob/ea68951b5949e77504fd8e5a829a7bcc722946b9/lib/vrl/compiler/src/precompiled/std/src/lib.rs) file using the [`-Z build-std=std`](https://github.com/vectordotdev/vector/blob/ea68951b5949e77504fd8e5a829a7bcc722946b9/lib/vrl/compiler/build.sh#L22) flag.